sdk/helper/keysutil/policy.go

```diff
@@ -1973,3 +1951,160 @@ func (p *Policy) EncryptWithFactory(ver int, context []byte, nonce []byte, value
+		rsaKey := parsedPrivateKey.(*rsa.PrivateKey)
+		if !rsaKey.PublicKey.Equal(keyEntry.RSAPublicKey) {
+			// NOTE: Description
+			return fmt.Errorf("keys do not match")
+		}
+	}
+
+	err = p.parseKeyToKeyEntry(&keyEntry, parsedPrivateKey, isPrivateKey)
+	if err != nil {
+		return err
+	}
+
+	p.Keys[strconv.Itoa(keyVersion)] = keyEntry
+
+	return p.Persist(ctx, storage)
+}
+
+// NOTE: func name?/not a fan of changing input params
+func (p *Policy) parseKeyToKeyEntry(entry *KeyEntry, parsedKey any, isPrivateKey bool) error {
+	switch parsedKey.(type) {
+	case *ecdsa.PrivateKey, *ecdsa.PublicKey:
+		if p.Type != KeyType_ECDSA_P256 && p.Type != KeyType_ECDSA_P384 && p.Type != KeyType_ECDSA_P521 {
+			return fmt.Errorf("invalid key type: expected %s, got %T", p.Type, parsedKey)
+		}
+
+		curve := elliptic.P256()
+		if p.Type == KeyType_ECDSA_P384 {
+			curve = elliptic.P384()
+		} else if p.Type == KeyType_ECDSA_P521 {
+			curve = elliptic.P521()
+		}
+
+		var derBytes []byte
+		var err error
+		if isPrivateKey {
+			ecdsaKey := parsedKey.(*ecdsa.PrivateKey)
+
+			if ecdsaKey.Curve != curve {
+				return fmt.Errorf("invalid curve: expected %s, got %s", curve.Params().Name, ecdsaKey.Curve.Params().Name)
+			}
+
+			entry.EC_D = ecdsaKey.D
+			entry.EC_X = ecdsaKey.X
+			entry.EC_Y = ecdsaKey.Y
+
+			derBytes, err = x509.MarshalPKIXPublicKey(ecdsaKey.Public())
+			if err != nil {
+				return errwrap.Wrapf("error marshaling public key: {{err}}", err)
+			}
+		} else {
+			ecdsaKey := parsedKey.(*ecdsa.PublicKey)
+
+			if ecdsaKey.Curve != curve {
+				return fmt.Errorf("invalid curve: expected %s, got %s", curve.Params().Name, ecdsaKey.Curve.Params().Name)
+			}
+
+			entry.EC_X = ecdsaKey.X
+			entry.EC_Y = ecdsaKey.Y
+
+			derBytes, err = x509.MarshalPKIXPublicKey(ecdsaKey)
+			if err != nil {
+				return errwrap.Wrapf("error marshaling public key: {{err}}", err)
+			}
+		}
+
+		pemBlock := &pem.Block{
+			Type:  "PUBLIC KEY",
+			Bytes: derBytes,
+		}
+		pemBytes := pem.EncodeToMemory(pemBlock)
+		if pemBytes == nil || len(pemBytes) == 0 {
+			return fmt.Errorf("error PEM-encoding public key")
+		}
+		entry.FormattedPublicKey = string(pemBytes)
+	case ed25519.PrivateKey:
+		if p.Type != KeyType_ED25519 {
+			return fmt.Errorf("invalid key type: expected %s, got %T", p.Type, parsedKey)
+		}
+		privateKey := parsedKey.(ed25519.PrivateKey)
+
+		entry.Key = privateKey
+		publicKey := privateKey.Public().(ed25519.PublicKey)
+		entry.FormattedPublicKey = base64.StdEncoding.EncodeToString(publicKey)
+	case *rsa.PrivateKey, *rsa.PublicKey:
+		if p.Type != KeyType_RSA2048 && p.Type != KeyType_RSA3072 && p.Type != KeyType_RSA4096 {
+			return fmt.Errorf("invalid key type: expected %s, got %T", p.Type, parsedKey)
+		}
+
+		keyBytes := 256
+		if p.Type == KeyType_RSA3072 {
+			keyBytes = 384
+		} else if p.Type == KeyType_RSA4096 {
+			keyBytes = 512
+		}
+		// NOTE: If for some reason we don't pass the `isPrivateKey`, we can use the keyTypes
```

Comment:
As an alternative if you like it, you could use the switch to dispatch correctly here.

But I think we _do_ want the enforcement between what we intend to do (at the higher level `p.Import...` and what we actually do (whether we import a public/private at the end of the day).

Reply:
Thanks for the feedback.
Not sure if I understood this comment but can be clarified in the PR to main.